lib/model/typing_status.dart

```diff
@@ -0,0 +1,83 @@
+import 'dart:async';
+
+import 'package:flutter/foundation.dart';
+
+import '../api/model/events.dart';
+import 'narrow.dart';
+
+/// The view-model for tracking the typing status organized by narrows.
+class TypingStatus extends ChangeNotifier {
+  @visibleForTesting
+  final Map<SendableNarrow, Map<int, Timer>> typistIdsByNarrow = {};
+
+  final int selfUserId;
+  final Duration typingStartedExpiryPeriod;
+  final Duration typingStoppedWaitPeriod;
+  final Duration typingStartedWaitPeriod;
+
+  TypingStatus({
+    required this.selfUserId,
+    required this.typingStartedExpiryPeriod,
+    required this.typingStoppedWaitPeriod,
+    required this.typingStartedWaitPeriod,
+  });
+
+  @override
+  void dispose() {
+    for (final typistTimer in typistIdsByNarrow.values) {
+      for (final timer in typistTimer.values) {
+        timer.cancel();
+      }
+    }
+    super.dispose();
+  }
+
+  bool _addTypist(SendableNarrow narrow, int typistUserId, void Function() callback) {
+    final narrowTimerMap = typistIdsByNarrow[narrow] ??= {};
+    final typistTimer = narrowTimerMap[typistUserId];
+    final isNewTypist = typistTimer == null;
+    typistTimer?.cancel();
+    narrowTimerMap[typistUserId] = Timer(typingStartedExpiryPeriod, callback);
+    return isNewTypist;
+  }
```

Comment:
The `callback` can just be inlined in here, can't it? It doesn't close over anything that we can't access here, right:

```suggestion
  bool _addTypist(SendableNarrow narrow, int typistUserId) {
    final narrowTimerMap = typistIdsByNarrow[narrow] ??= {};
    final typistTimer = narrowTimerMap[typistUserId];
    final isNewTypist = typistTimer == null;
    typistTimer?.cancel();
    narrowTimerMap[typistUserId] = Timer(typingStartedExpiryPeriod, () {
      if (_removeTypist(narrow, typistUserId)) {
        notifyListeners();
      }
    });
    return isNewTypist;
  }
```

Reply:
Oh, and since this `notifyListeners` happens after a delay, let's put it behind a check that the instance hasn't been `dispose`d yet. I'm not sure if there's a better way to check that than just `hasListeners`.

Reply:
We can also track it with a `bool`, but `hasListeners` should be enough.